src/block/CaveVines.php

```diff
@@ -0,0 +1,200 @@
+	}
+
+	public function getAge() : int{
+		return $this->age;
+	}
+
+	/** @return $this */
+	public function setAge(int $age) : self{
+		if($age < 0 || $age > self::MAX_AGE){
+			throw new \InvalidArgumentException("Age must be in range 0-" . self::MAX_AGE);
+		}
+		$this->age = $age;
+		return $this;
+	}
+
+	public function hasBerries() : bool{
+		return $this->hasBerries;
+	}
+
+	/** @return $this */
+	public function setBerries(bool $hasBerries) : self{
+		$this->hasBerries = $hasBerries;
+		return $this;
+	}
+
+	public function isTip() : bool{
+		return $this->tip;
+	}
+
+	/** @return $this */
+	public function setTip(bool $tip) : self{
+		$this->tip = $tip;
+		return $this;
+	}
+
+	public function isAffectedBySilkTouch() : bool{
+		return true;
+	}
+
+	public function ticksRandomly() : bool{
+		return true;
+	}
+
+	public function canClimb() : bool{
+		return true;
+	}
+
+	public function getLightLevel() : int{
+		return $this->hasBerries ? 14 : 0;
+	}
+
+	private function canBeSupportedBy(Block $block) : bool{
+		return $block->getSupportType(Facing::DOWN)->hasCenterSupport() || $block->isSameType($this);
+	}
+
+	public function onNearbyBlockChange() : void{
+		if(!$this->canBeSupportedBy($this->getSide(Facing::UP))){
+			$this->position->getWorld()->useBreakOn($this->position);
+		}
+	}
+
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if(!$this->canBeSupportedBy($blockReplace->getSide(Facing::UP))){
+			return false;
+		}
+		$this->age = mt_rand(0, self::MAX_AGE);
+		return parent::place($tx, $item, $blockReplace, $blockClicked, $face, $clickVector, $player);
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($this->hasBerries){
+			$this->position->getWorld()->dropItem($this->position, VanillaItems::GLOW_BERRIES());
+			$this->position->getWorld()->addSound($this->position, new GlowBerriesPickSound());
+
+			$this->hasBerries = false;
+			$this->age = mt_rand(0, self::MAX_AGE);
+			$this->position->getWorld()->setBlock($this->position, $this);
+			return true;
+		}
+		if($item instanceof Fertilizer){
+			$ev = new BlockGrowEvent($this, (clone $this)->setBerries(true));
+			$ev->call();
+			if($ev->isCancelled()){
+				return false;
+			}
+			$item->pop();
+			$this->position->getWorld()->setBlock($this->position, $ev->getNewState());
+			return true;
+		}
+		return false;
+	}
+
+	public function onRandomTick() : void{
+		if($this->tip && $this->age < self::MAX_AGE && mt_rand(1, 10) === 1){
+			$growthPos = $this->position->getSide(Facing::DOWN);
+			$world = $growthPos->getWorld();
+			if($world->isInWorld($growthPos->getFloorX(), $growthPos->getFloorY(), $growthPos->getFloorZ()) && $world->getBlock($growthPos)->getTypeId() === BlockTypeIds::AIR){
+				$ev = new BlockGrowEvent($this, VanillaBlocks::CAVE_VINES()
```

Comment:
This position doesn't match the position used below.

Reply:
The first argument to `BlockGrowEvent` should be the block which is being replaced. This event incorrectly suggests that $this is being replaced.